concepts/00001-enarx-sprint-framework/README.md

```diff
@@ -0,0 +1,158 @@
+
+### Backlog
+
+As our project scales, it’s easy to imagine a single unified backlog becoming incredibly cumbersome and hard to manage. Instead, we should adapt our current approach to area-specific boards, which effectively divides our backlog into categories (SGX, SEV,  infrastructure, etc.).
+
+Each area-specific backlog will be housed in a Github project with the following columns:
+
+- Backlog: for issues we want to tackle eventually, but not a high priority
+- Tracking: for issues that are high-priority, and that we want to take action on sooner rather than later. All new project issues should land here by default to ensure they don’t get lost (automatable with stock Github).
+- Active: for issues that are actively being worked on as part of the current sprint.
+- Done: when things are done, they go here (automatable with stock Github). Mostly to ensure the “Active” column stays organized.
+
+In addition to inherently categorizing our backlog, this approach allows new "categories" to be moved around or created as needed -- and, if we decide to use organization-level projects for our sprints, allows the backlog to be bisected across repositories if wanted. This will support modularization as the project grows.
+
+To start, I propose the following backlog projects:
+
+- Documentation
+- Infrastructure
+- SEV
+- SGX
+- WASM
+
+A few things to consider:
+- Downside: issues that don’t fit squarely into a category don’t have a place to go. Maybe a "Miscellaneous" project?
+- Automation opportunity: we could automatically move issues from Tracking to Active when they are assigned to someone. Automatable with Actions.
+
+### Active Sprint Board
+This board is where the team will track progress on current sprint tasks. **It should hold no backlog;** if an issue or PR is on this board, it is being actively worked on.
+
+Here’s how the columns should look:
+
+- Assigned: this indicates a task that is currently being worked on as part of the sprint. Newly-added issues land here by default (automatable with stock Github).
+- Ready for review: where pull requests land by default (automatable with stock Github); indicates it’s ready to be reviewed by someone. Pull requests should move back here from Changes Requested if a fix has been applied and it’s ready for a re-review (not in stock Github, but automatable with Actions).
+- Changes Requested: PRs move here when a review is submitted that requests changes (automatable with stock Github).
+- Approved: PRs that have been approved go here, pending merge (automatable with stock Github).
+- Done: PRs that have been merged, and their associated issues, go here (automatable with stock Github).
+
+A few things to consider:
+- Advantage: having project board be separate from backlog means we can promote issues to be part of the current sprint whenever we want to, and they can skip the backlog entirely if that’s desired.
```

Comment:
I think we should be cautious with a mechanism like this. If we abuse it then it undermines the efforts that go into sprint planning.

Reply:
@connorkuehl Agreed. However, I asked for this explicitly because I want a way to capture drive-by contributions. I don't want a new contributor to have to wait 2 weeks for a PR just because they missed the planning meeting by 5 minutes.

Reply:
I also have it come up pretty often that something I'm working on (or something that comes up within @npmccallum 's work) spawns a new issue that gets promoted to my immediate workload... so not being able to do that seems strange? Cumbersome? I can think of sprint-compliant ways to do this sort of thing, but we should probably talk about what we want that to look like.

Reply:
> Agreed. However, I asked for this explicitly because I want a way to capture drive-by contributions. I don't want a new contributor to have to wait 2 weeks for a PR just because they missed the planning meeting by 5 minutes.

This makes sense, but i think drive-by contributors should not be considered to be "part of the team", and as such exempt from following sprints.

@lkatalin According to agile doctrine, this would suggest that the original issue was either not groomed properly, or not broken down into the right subtasks during sprint planning. The spawned issue you run into would likely have emerged as the issue-owner tries to explain the issue to the rest of the team and everyone reaches a common understanding (backlog grooming) or as the issue is broken down into immediately actionable tasks (sprint planning).

This is all a bit theoretical of course, because there's always stuff that just wasn't foreseeable. But then again, the concept of spikes, or stories to explore something that isn't well understood enough to be done as a normal user story exists for this case. :smiley:

Reply:
@axelsimon Yeah, I think this partially speaks to my (mild) skepticism that some of our workflow can entirely fit into the Platonic ideal "sprint" because it's all new ground and we often don't know what the sub-tasks or blockers are until we try a few different things. But it's also possible that we can be more exact about sub-tasks than we currently are. Only time will tell. I am okay with trying to conform to the sprint and seeing how it goes, and adapting it as required.

Reply:
As I noted on the call today, I think the right way to move forward here is to treat issues and PRs differently. Sprint planning decides which issues are placed on the sprint board. Issues may or may not have a corresponding PR. All PRs go on the sprint board, regardless of whether they have a corresponding issue or not.

This has the following properties.

1. We prioritize the work of the core team.
2. We can have non-code issues as part of planning. For example, infrastructure issues. These don't need PRs.
3. All PRs have the same process visibility. Whether your work is "core team" work or "outside contributor work" you get the same visibility.
4. Joining the "core team" means you take part in sprint planning. This is open to anyone.

Reply:
One thing I forgot to mention two things regarding issues.

1. We probably need some sort of issue triage that is more frequent than the Sprint duration. At the very least we want a prompt human touch on each issue that is filed so that contributors feel welcomed. We also need to identify urgent issues...

2. We need to have a notion of "interrupt" issues which bypass sprint planning and go directly to the board even in the middle of a sprint. These should be rare. They should either (1) block existing sprint work or (2) have otherwise catastrophic impact (think: website is down).

3. What is the process for mid-sprint decomposing? What if we have planned an issue for a sprint and while doing it we notice it should be decomposed further? How do we handle this case?

Reply:
I second that mid-sprint decomposing or some type of check-in re: the chosen work would be beneficial.

Reply:
> 3. What is the process for mid-sprint decomposing? What if we have planned an issue for a sprint and while doing it we notice it should be decomposed further? How do we handle this case?

Someone needs to make a call (or a specified group of people, or a process for voting, etc.).  This is another reason that we need to define roles and responsibilities.

Reply:
@npmccallum My thinking around point 1) of yours above was that we'd do new issue triage _every standup_. This would mean new issues get a human response within 24 working hours. Were you thinking of something in addition to that?

Reply:
No.